<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>preciseTAD: A machine learning framework for precise TAD boundary 
    prediction • preciseTAD</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="preciseTAD: A machine learning framework for precise TAD boundary 
    prediction">
<meta property="og:description" content="preciseTAD provides functions to predict the location of boundaries 
    of topologically associated domains (TADs) and chromatin loops at base-level 
    resolution. As an input, it takes BED-formatted genomic coordinates of 
    domain boundaries detected from low-resolution Hi-C data, and coordinates of 
    high-resolution genomic annotations from ENCODE or other consortia. 
    preciseTAD employs several feature engineering strategies and resampling 
    techniques to address class imbalance, and trains an optimized random forest 
    model for predicting low-resolution domain boundaries. Translated on 
    a base-level, preciseTAD predicts the probability for each base to be 
    a boundary. Density-based clustering and scalable partitioning techniques 
    are used to detect precise boundary regions and summit points. Compared with 
    low-resolution boundaries, preciseTAD boundaries are highly enriched for 
    CTCF, RAD21, SMC3, and ZNF143 signal and more conserved across cell lines. 
    The pre-trained model can accurately predict boundaries in another cell line 
    using CTCF, RAD21, SMC3, and ZNF143 annotation data for this cell line. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">preciseTAD</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/preciseTAD.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dozmorovlab/preciseTAD/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="precisetad-a-transfer-learning-framework-for-3d-domain-boundary-prediction-at-base-pair-resolution" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#precisetad-a-transfer-learning-framework-for-3d-domain-boundary-prediction-at-base-pair-resolution" class="anchor"></a>preciseTAD: A transfer learning framework for 3D domain boundary prediction at base-pair resolution</h1></div>
<!-- [![Travis build
status](https://travis-ci.com/stilianoudakis/preciseTAD.svg?branch=master)](https://travis-ci.com/stilianoudakis/preciseTAD) -->
<!-- badges: start -->

<p>Stilianoudakis, Spiro, and Mikhail G. Dozmorov. “preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution.” bioRxiv (2020). <a href="https://doi.org/10.1101/2020.09.03.282186" class="uri">https://doi.org/10.1101/2020.09.03.282186</a></p>
<p>Predicted preciseTAD boundary points (PTBPs) and regions (PTBRs) for 60 cell lines are available <a href="https://drive.google.com/drive/folders/15Rc6PhrrBjThwE-5dSyNX-ILELaUu6uG?usp=sharing">here</a>.</p>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>preciseTAD provides functions to predict the location of boundaries of topologically associated domains (TADs) and chromatin loops at base-level resolution. As an input, it takes BED-formatted genomic coordinates of domain boundaries detected from low-resolution Hi-C data, and coordinates of high-resolution genomic annotations from ENCODE or other consortia. preciseTAD employs several feature engineering strategies and resampling techniques to address class imbalance, and trains an optimized random forest model for predicting low-resolution domain boundaries. Translated on a base-level, preciseTAD predicts the probability for each base to be a boundary. Density-based clustering and scalable partitioning techniques are used to detect precise boundary regions and summit points. Compared with low-resolution boundaries, preciseTAD boundaries are highly enriched for CTCF, RAD21, SMC3, and ZNF143 signal and more conserved across cell lines. The pre-trained model can accurately predict boundaries in another cell line using CTCF, RAD21, SMC3, and ZNF143 annotation data for this cell line.</p>
<p>The main functions (in order of implementation) are:</p>
<ul>
<li>
<code><a href="reference/extractBoundaries.html">extractBoundaries()</a></code> accepts a 3-column data.frame or matrix with the chromosomal coordinates of user-defined domains and outputs the unique boundaries. The second and third columns are the domain anchor centers.</li>
<li>
<code><a href="reference/bedToGRangesList.html">bedToGRangesList()</a></code> accepts a filepath containing BED files representing the coordinates of ChIP-seq defined functional genomic annotations</li>
<li>
<code><a href="reference/createTADdata.html">createTADdata()</a></code> accepts a set of unique boundaries and genomic annotations derived from <code><a href="reference/extractBoundaries.html">extractBoundaries()</a></code> and <code><a href="reference/bedToGRangesList.html">bedToGRangesList()</a></code>, respectively, to create the data matrix used to build a model to predict domain boundary regions</li>
<li>
<code><a href="reference/TADrandomForest.html">TADrandomForest()</a></code> a wrapper of the <code>randomForest</code> package which implements a random forest binary classification algorithm on domain boundary data</li>
<li>
<code><a href="reference/preciseTAD.html">preciseTAD()</a></code> which leverages a domain boundary prediction model (i.e., random forest) and density-based clustering to predict TAD boundary coordinates at a base-level resolution</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><code>preciseTAD</code> can be installed from Bioconductor:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># if (!requireNamespace("BiocManager", quietly=TRUE))</span>
<span class="co">#     install.packages("BiocManager")</span>
<span class="co"># BiocManager::install("preciseTAD")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dozmorovlab/preciseTAD">preciseTAD</a></span><span class="op">)</span>
<span class="co">#&gt; </span></code></pre></div>
<p>The latest version of <code>preciseTAD</code> can be directly installed from Github:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"dozmorovlab/preciseTAD"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dozmorovlab/preciseTAD">preciseTAD</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>Below is a brief workflow of how to implement <code>preciseTAD</code> on binned data from CHR1 to get precise base pair coordinates of TAD boundaries for a 10mb section of CHR 22. For more details, including the example how to use the pre-trained model, see <code><a href="articles/preciseTAD.html">vignette("preciseTAD")</a></code></p>
<p>First, you need to obtain called TAD boundaries using an established TAD-caller. As an example, consider the <a href="https://github.com/aidenlab/juicer/wiki/Arrowhead">Arrowhead</a> TAD-caller, a part of the juicer suite of tools developed by the Aiden Lab. Arrowhead outputs a .txt file with the chromosomal start and end coordinates of their called TADs. As an example, we have provided Arrowhead TADs for GM12878 at 5kb resolution.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"arrowhead_gm12878_5kb"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">arrowhead_gm12878_5kb</span><span class="op">)</span>
<span class="co">#&gt;   V1        V2        V3</span>
<span class="co">#&gt; 1  1  49375000  50805000</span>
<span class="co">#&gt; 2  1  16830000  17230000</span>
<span class="co">#&gt; 3  1 163355000 164860000</span>
<span class="co">#&gt; 4  1 231935000 233400000</span>
<span class="co">#&gt; 5  1 149035000 149430000</span>
<span class="co">#&gt; 6  1   3995000   5505000</span></code></pre></div>
<p>The unique boundaries for CHR1 and CHR22 can be extracted as:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/extractBoundaries.html">extractBoundaries</a></span><span class="op">(</span>domains.mat <span class="op">=</span> <span class="va">arrowhead_gm12878_5kb</span>, filter <span class="op">=</span> <span class="cn">FALSE</span>, CHR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CHR1"</span>, <span class="st">"CHR22"</span><span class="op">)</span>, resolution <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span>
<span class="va">bounds</span>
<span class="co">#&gt; GRanges object with 1901 ranges and 0 metadata columns:</span>
<span class="co">#&gt;         seqnames            ranges strand</span>
<span class="co">#&gt;            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;     787     chr1     815000-815001      *</span>
<span class="co">#&gt;    9196     chr1     890000-890001      *</span>
<span class="co">#&gt;      37     chr1     915000-915001      *</span>
<span class="co">#&gt;    8923     chr1   1005000-1005001      *</span>
<span class="co">#&gt;     275     chr1   1015000-1015001      *</span>
<span class="co">#&gt;     ...      ...               ...    ...</span>
<span class="co">#&gt;    8379    chr22 50815000-50815001      *</span>
<span class="co">#&gt;   16788    chr22 50935000-50935001      *</span>
<span class="co">#&gt;    8382    chr22 50945000-50945001      *</span>
<span class="co">#&gt;   16791    chr22 51060000-51060001      *</span>
<span class="co">#&gt;   16670    chr22 51235000-51235001      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 22 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
<p>Next, you will need to download cell line-specific ChIP-seq data in the form of BED files from <a href="https://www.encodeproject.org/chip-seq-matrix/?type=Experiment&amp;replicates.library.biosample.donor.organism.scientific_name=Homo%20sapiens&amp;assay_title=TF%20ChIP-seq&amp;status=released">ENCODE</a>. Once, you have downloaded your preferred list of functional genomic annotations, store them in a specific file location. These files can then be converted into a GRangesList object and used for downstream modeling using the following command:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"pathToBEDfiles"</span>
<span class="va">tfbsList</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bedToGRangesList.html">bedToGRangesList</a></span><span class="op">(</span>filepath <span class="op">=</span> <span class="va">path</span>, bedList <span class="op">=</span> <span class="cn">NULL</span>, bedNames <span class="op">=</span> <span class="cn">NULL</span>, pattern <span class="op">=</span> <span class="st">"*.bed"</span>, signal <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></code></pre></div>
<p>As an example, we have already provided a GRangesList object with a variety of transcription factor binding sites specific to the GM12878 cell line. Once you load it in, you can see the list of transcription factors using the following:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"tfbsList"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tfbsList</span><span class="op">)</span>
<span class="co">#&gt;  [1] "Gm12878-Atf3-Haib"       "Gm12878-Cfos-Sydh"      </span>
<span class="co">#&gt;  [3] "Gm12878-Cmyc-Uta"        "Gm12878-Ctcf-Broad"     </span>
<span class="co">#&gt;  [5] "Gm12878-Egr1-Haib"       "Gm12878-Ets1-Haib"      </span>
<span class="co">#&gt;  [7] "Gm12878-Gabp-Haib"       "Gm12878-Jund-Sydh"      </span>
<span class="co">#&gt;  [9] "Gm12878-Max-Sydh"        "Gm12878-Mazab85725-Sydh"</span>
<span class="co">#&gt; [11] "Gm12878-Mef2a-Haib"      "Gm12878-Mxi1-Sydh"      </span>
<span class="co">#&gt; [13] "Gm12878-P300-Sydh"       "Gm12878-Pol2-Haib"      </span>
<span class="co">#&gt; [15] "Gm12878-Pu1-Haib"        "Gm12878-Rad21-Haib"     </span>
<span class="co">#&gt; [17] "Gm12878-Rfx5-Sydh"       "Gm12878-Sin3a-Sydh"     </span>
<span class="co">#&gt; [19] "Gm12878-Six5-Haib"       "Gm12878-Smc3-Sydh"      </span>
<span class="co">#&gt; [21] "Gm12878-Sp1-Haib"        "Gm12878-Srf-Haib"       </span>
<span class="co">#&gt; [23] "Gm12878-Taf1-Haib"       "Gm12878-Tr4-Sydh"       </span>
<span class="co">#&gt; [25] "Gm12878-Yy1-Sydh"        "Gm12878-Znf143-Sydh"</span></code></pre></div>
<p>For the purposes of this example, let’s focus only on CTCF, RAD21, SMC3, and ZNF143 transcription factors.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tfbsList_filt</span> <span class="op">&lt;-</span> <span class="va">tfbsList</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tfbsList</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gm12878-Ctcf-Broad"</span>, <span class="st">"Gm12878-Rad21-Haib"</span>, <span class="st">"Gm12878-Smc3-Sydh"</span>, <span class="st">"Gm12878-Znf143-Sydh"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>Now, using the “ground-truth” boundaries and the following TFBS, we can build the data matrix that will be used for predictive modeling. The following command creates the training data from CHR1 and reserves the testing data from CHR22. We specify 5kb sized genomic bins (to match the resolution used to call the original TADs), a distance-type feature space, and apply random under-sampling (RUS) on the training data only.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">tadData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/createTADdata.html">createTADdata</a></span><span class="op">(</span>bounds.GR          <span class="op">=</span> <span class="va">bounds</span>,
                         resolution         <span class="op">=</span> <span class="fl">5000</span>,
                         genomicElements.GR <span class="op">=</span> <span class="va">tfbsList_filt</span>,
                         featureType        <span class="op">=</span> <span class="st">"distance"</span>,
                         resampling         <span class="op">=</span> <span class="st">"rus"</span>,
                         trainCHR           <span class="op">=</span> <span class="st">"CHR1"</span>,
                         predictCHR         <span class="op">=</span> <span class="st">"CHR22"</span>
<span class="op">)</span></code></pre></div>
<p>We can now implement our machine learning algorithm of choice to predict TAD-boundary regions. Here, we opt for the random forest algorithm.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">tadModel</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TADrandomForest.html">TADrandomForest</a></span><span class="op">(</span>trainData    <span class="op">=</span> <span class="va">tadData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                            testData     <span class="op">=</span> <span class="va">tadData</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,
                            tuneParams   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mtry <span class="op">=</span> <span class="fl">2</span>, ntree <span class="op">=</span> <span class="fl">500</span>, nodesize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
                            cvFolds      <span class="op">=</span> <span class="fl">3</span>,
                            cvMetric     <span class="op">=</span> <span class="st">"Accuracy"</span>,
                            verbose      <span class="op">=</span> <span class="cn">TRUE</span>,
                            model        <span class="op">=</span> <span class="cn">TRUE</span>,
                            importances  <span class="op">=</span> <span class="cn">TRUE</span>,
                            impMeasure   <span class="op">=</span> <span class="st">"MDA"</span>,
                            performances <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: lattice</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="co">#&gt; + Fold1: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold1: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; + Fold2: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold2: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; + Fold3: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold3: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; Aggregating results</span>
<span class="co">#&gt; Fitting final model on full training set</span>
<span class="co"># The model itself</span>
<span class="va">tadModel</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; Random Forest </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 3190 samples</span>
<span class="co">#&gt;    4 predictor</span>
<span class="co">#&gt;    2 classes: 'No', 'Yes' </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; No pre-processing</span>
<span class="co">#&gt; Resampling: Cross-Validated (3 fold) </span>
<span class="co">#&gt; Summary of sample sizes: 2126, 2127, 2127 </span>
<span class="co">#&gt; Resampling results:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   MCC        ROC        Sens    Spec       Pos Pred Value  Neg Pred Value</span>
<span class="co">#&gt;   0.4679235  0.7990197  0.7072  0.7598457  0.747661        0.7224703     </span>
<span class="co">#&gt;   Accuracy   Kappa    </span>
<span class="co">#&gt;   0.7335413  0.4670627</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Tuning parameter 'mtry' was held constant at a value of 2</span>
<span class="co">#&gt; Tuning</span>
<span class="co">#&gt;  parameter 'ntree' was held constant at a value of 500</span>
<span class="co">#&gt; Tuning</span>
<span class="co">#&gt;  parameter 'nodesize' was held constant at a value of 1</span>

<span class="co"># Variable importances (mean decrease in accuracy)</span>
<span class="va">tadModel</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt;                 Feature Importance</span>
<span class="co">#&gt; 4 `Gm12878-Znf143-Sydh`   75.98200</span>
<span class="co">#&gt; 3   `Gm12878-Smc3-Sydh`   62.50681</span>
<span class="co">#&gt; 2  `Gm12878-Rad21-Haib`   62.21715</span>
<span class="co">#&gt; 1  `Gm12878-Ctcf-Broad`   32.61855</span>

<span class="co"># Model performance metrics</span>
<span class="va">tadModel</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt;              Metric  Performance</span>
<span class="co">#&gt; 1                TN 6.400000e+03</span>
<span class="co">#&gt; 2                FN 6.700000e+01</span>
<span class="co">#&gt; 3                FP 2.954000e+03</span>
<span class="co">#&gt; 4                TP 2.390000e+02</span>
<span class="co">#&gt; 5             Total 9.660000e+03</span>
<span class="co">#&gt; 6       Sensitivity 7.810458e-01</span>
<span class="co">#&gt; 7       Specificity 6.841993e-01</span>
<span class="co">#&gt; 8             Kappa 8.363202e-02</span>
<span class="co">#&gt; 9          Accuracy 6.872671e-01</span>
<span class="co">#&gt; 10 BalancedAccuracy 7.326225e-01</span>
<span class="co">#&gt; 11        Precision 7.485124e-02</span>
<span class="co">#&gt; 12              FPR 3.158007e-01</span>
<span class="co">#&gt; 13              FNR 1.036029e-02</span>
<span class="co">#&gt; 14              NPV 9.896397e-01</span>
<span class="co">#&gt; 15              MCC 1.732169e-01</span>
<span class="co">#&gt; 16               F1 1.366105e-01</span>
<span class="co">#&gt; 17              AUC 7.983352e-01</span>
<span class="co">#&gt; 18           Youden 4.652450e-01</span>
<span class="co">#&gt; 19            AUPRC 9.383824e-02</span></code></pre></div>
<p>Lastly, we take our TAD-boundary region predictive model and use it to make predictions on a 10mb section of CHR22:35,000,000-45,000,000.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run preciseTAD</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">pt</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/preciseTAD.html">preciseTAD</a></span><span class="op">(</span> genomicElements.GR <span class="op">=</span> <span class="va">tfbsList_filt</span>,
                  featureType        <span class="op">=</span> <span class="st">"distance"</span>,
                  CHR                <span class="op">=</span> <span class="st">"CHR22"</span>,
                  chromCoords        <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fl">35000000</span>, <span class="fl">45000000</span><span class="op">)</span>,
                  tadModel           <span class="op">=</span> <span class="va">tadModel</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                  threshold          <span class="op">=</span> <span class="fl">1.0</span>,
                  verbose            <span class="op">=</span> <span class="cn">FALSE</span>,
                  parallel           <span class="op">=</span> <span class="fl">2</span>,
                  DBSCAN_params      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fl">30000</span>, <span class="fl">3</span><span class="op">)</span>,
                  slope              <span class="op">=</span> <span class="fl">5000</span>,
                  genome             <span class="op">=</span> <span class="st">"hg19"</span><span class="op">)</span>

<span class="co"># View preciseTAD predicted boundary coordinates between CHR22:35mb-45mb</span>
<span class="va">pt</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; GRanges object with 64 ranges and 0 metadata columns:</span>
<span class="co">#&gt;        seqnames            ranges strand</span>
<span class="co">#&gt;           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;    [1]    chr22 35337875-35342203      *</span>
<span class="co">#&gt;    [2]    chr22 35413232-35426376      *</span>
<span class="co">#&gt;    [3]    chr22 35543040-35549073      *</span>
<span class="co">#&gt;    [4]    chr22 35620677-35634681      *</span>
<span class="co">#&gt;    [5]    chr22 35740783-35747642      *</span>
<span class="co">#&gt;    ...      ...               ...    ...</span>
<span class="co">#&gt;   [60]    chr22 43261091-43270096      *</span>
<span class="co">#&gt;   [61]    chr22 43425520-43490629      *</span>
<span class="co">#&gt;   [62]    chr22 43774902-43791526      *</span>
<span class="co">#&gt;   [63]    chr22 44272171-44283944      *</span>
<span class="co">#&gt;   [64]    chr22 44382230-44396240      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/preciseTAD">https://​www.bioconductor.org/​packages/​preciseTAD</a>
</li>
<li>Browse source code at <br><a href="https://github.com/dozmorovlab/preciseTAD/">https://​github.com/​dozmorovlab/​preciseTAD/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/dozmorovlab/preciseTAD/issues">https://​github.com/​dozmorovlab/​preciseTAD/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Spiro Stilianoudakis <br><small class="roles"> Author, maintainer </small>  </li>
<li>Mikhail Dozmorov <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle: maturing"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Spiro Stilianoudakis, Mikhail Dozmorov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
