<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function to create a data matrix used for building a predictive model to
classify boundary regions from functional genomic elements — createTADdata • preciseTAD</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function to create a data matrix used for building a predictive model to
classify boundary regions from functional genomic elements — createTADdata" />
<meta property="og:description" content="Function to create a data matrix used for building a predictive model to
classify boundary regions from functional genomic elements" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">preciseTAD</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/preciseTAD.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dozmorovlab/preciseTAD/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Function to create a data matrix used for building a predictive model to
classify boundary regions from functional genomic elements</h1>
    <small class="dont-index">Source: <a href='https://github.com/dozmorovlab/preciseTAD/blob/master/R/createTADdata.R'><code>R/createTADdata.R</code></a></small>
    <div class="hidden name"><code>createTADdata.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Function to create a data matrix used for building a predictive model to
classify boundary regions from functional genomic elements</p>
    </div>

    <pre class="usage"><span class='fu'>createTADdata</span><span class='op'>(</span>
  <span class='va'>bounds.GR</span>,
  <span class='va'>resolution</span>,
  <span class='va'>genomicElements.GR</span>,
  featureType <span class='op'>=</span> <span class='st'>"distance"</span>,
  <span class='va'>resampling</span>,
  <span class='va'>trainCHR</span>,
  predictCHR <span class='op'>=</span> <span class='cn'>NULL</span>,
  genome <span class='op'>=</span> <span class='st'>"hg19"</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bounds.GR</th>
      <td><p>a GRanges object with chromosomal coordinates of TAD
boundaries used to identify positive cases (can be obtained using
<code><a href='extractBoundaries.html'>extractBoundaries</a></code>). Required.</p></td>
    </tr>
    <tr>
      <th>resolution</th>
      <td><p>Numeric, the width to bin the genome at, should match the
resolution that TADs were called at. Required.</p></td>
    </tr>
    <tr>
      <th>genomicElements.GR</th>
      <td><p>a GRangesList object containing GRanges objects
for each ChIP-seq data to leverage in the random forest model (can be
obtained using the <code><a href='bedToGRangesList.html'>bedToGRangesList</a></code>). Required.</p></td>
    </tr>
    <tr>
      <th>featureType</th>
      <td><p>Character, controls how the feature space is constructed
(one of either "binary" (overlap yes/no), "oc" (overlap counts, the number
of overlaps), "op" (overlap percent, the percent of bin width covered by the
genomic annotation), or "distance" (log2-transformed distance from the center
of the nearest genomic annotation to the center of the bin); default is
"distance"). Required.</p></td>
    </tr>
    <tr>
      <th>resampling</th>
      <td><p>Character, controls if and how the data should be
resampled to create balanced classes of boundary vs. nonboundary regions (one
of either "none" - no re-sampling, "ros" - Random Over-Sampling, "rus" -
Random Under-Sampling. Required.</p></td>
    </tr>
    <tr>
      <th>trainCHR</th>
      <td><p>Character vector of chromosomes to use to build the binned
data matrix for training. Required.</p></td>
    </tr>
    <tr>
      <th>predictCHR</th>
      <td><p>Character vector of chromosomes to use to build the binned
data matrix for testing. Default in NULL, indicating no test data is created.
 If trainCHR=predictCHR then a 7:3 split is created.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>version of the human genome assembly. Used to filter out
bases overlapping centromeric regions. Accepted values - hg19 (default) or 
hg38.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list object containing two data.frames: 1) the training data, 2)
the test data (only if predictCHR is not NULL, otherwise it is NA). "y" is
an indicator whether the corresponding bin is a TAD boundary, and the
subsequent columns have the association measures between bins and the
genomic annotations</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Create training data for CHR21 and testing data for CHR22 with</span>
<span class='co'># 5 kb binning, oc-type predictors from 26 different transcription factor</span>
<span class='co'># binding sites from the GM12878 cell line, and random under-sampling</span>

<span class='co'># Read in ARROWHEAD-called TADs at 5kb</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>arrowhead_gm12878_5kb</span><span class='op'>)</span>

<span class='co'>#Extract unique boundaries</span>
<span class='va'>bounds.GR</span> <span class='op'>&lt;-</span> <span class='fu'><a href='extractBoundaries.html'>extractBoundaries</a></span><span class='op'>(</span>domains.mat <span class='op'>=</span> <span class='va'>arrowhead_gm12878_5kb</span>,
                               filter <span class='op'>=</span> <span class='cn'>FALSE</span>,
                               CHR <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CHR21"</span>, <span class='st'>"CHR22"</span><span class='op'>)</span>,
                               resolution <span class='op'>=</span> <span class='fl'>5000</span><span class='op'>)</span>

<span class='co'># Read in GRangesList of 26 TFBS</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>tfbsList</span><span class='op'>)</span>

<span class='va'>tadData</span> <span class='op'>&lt;-</span> <span class='fu'>createTADdata</span><span class='op'>(</span>bounds.GR <span class='op'>=</span> <span class='va'>bounds.GR</span>,
                         resolution <span class='op'>=</span> <span class='fl'>5000</span>,
                         genomicElements.GR <span class='op'>=</span> <span class='va'>tfbsList</span>,
                         featureType <span class='op'>=</span> <span class='st'>"oc"</span>,
                         resampling <span class='op'>=</span> <span class='st'>"rus"</span>,
                         trainCHR <span class='op'>=</span> <span class='st'>"CHR21"</span>,
                         predictCHR <span class='op'>=</span> <span class='st'>"CHR22"</span><span class='op'>)</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Spiro Stilianoudakis, Mikhail Dozmorov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


