<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Precise TAD boundary prediction at base-level resolution using density-based
spatial clustering and partitioning techniques — preciseTAD • preciseTAD</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Precise TAD boundary prediction at base-level resolution using density-based
spatial clustering and partitioning techniques — preciseTAD" />
<meta property="og:description" content="Precise TAD boundary prediction at base-level resolution using density-based
spatial clustering and partitioning techniques" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">preciseTAD</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/preciseTAD.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dozmorovlab/preciseTAD/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Precise TAD boundary prediction at base-level resolution using density-based
spatial clustering and partitioning techniques</h1>
    <small class="dont-index">Source: <a href='https://github.com/dozmorovlab/preciseTAD/blob/master/R/preciseTAD.R'><code>R/preciseTAD.R</code></a></small>
    <div class="hidden name"><code>preciseTAD.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Precise TAD boundary prediction at base-level resolution using density-based
spatial clustering and partitioning techniques</p>
    </div>

    <pre class="usage"><span class='fu'>preciseTAD</span><span class='op'>(</span>
  <span class='va'>genomicElements.GR</span>,
  featureType <span class='op'>=</span> <span class='st'>"distance"</span>,
  <span class='va'>CHR</span>,
  chromCoords <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>tadModel</span>,
  threshold <span class='op'>=</span> <span class='fl'>1</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  parallel <span class='op'>=</span> <span class='cn'>NULL</span>,
  DBSCAN_params <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fl'>30000</span>, <span class='fl'>100</span><span class='op'>)</span>,
  slope <span class='op'>=</span> <span class='fl'>5000</span>,
  genome <span class='op'>=</span> <span class='st'>"hg19"</span>,
  BaseProbs <span class='op'>=</span> <span class='cn'>FALSE</span>,
  savetobed <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>genomicElements.GR</th>
      <td><p><code>GRangesList</code> object containing GRanges from
each ChIP-seq BED file that was used to train a predictive model (can be
obtained using the <code><a href='bedToGRangesList.html'>bedToGRangesList</a></code>). Required.</p></td>
    </tr>
    <tr>
      <th>featureType</th>
      <td><p>Controls how the feature space is constructed (one of
either "binary", "oc", "op", "signal, or "distance" (log2- transformed).
Default and recommended: "distance".</p></td>
    </tr>
    <tr>
      <th>CHR</th>
      <td><p>Controls which chromosome to predict boundaries on at base-level
resolution, e.g., CHR22. Required.</p></td>
    </tr>
    <tr>
      <th>chromCoords</th>
      <td><p>List containing the starting bp coordinate and ending bp
coordinate that defines the region of the linear genome to make predictions
on. If chromCoords is not specified, then predictions will be made on the
entire chromosome. Default is NULL.</p></td>
    </tr>
    <tr>
      <th>tadModel</th>
      <td><p>Model object used to obtain predicted probabilities at
base-level resolution (examples include <code>gbm</code>, <code>glmnet</code>,
<code>svm</code>, <code>glm</code>, etc). For a random forest model, can be obtained
using <code>preciseTAD::randomForest</code>). Required.</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>Bases with predicted probabilities that are greater
than or equal to this value are labeled as potential TAD boundaries. Values
in the range of .95-1.0 are suggested. Default is 1. To explore how
selection of the `threshold` parameter affects the results, it is recommended
to rerun the function with a different threshold, e.g., 0.99, and compare the
results of Normalized Enrichment test (see `DBSCAN_params` and the 
`preciseTADparams` slot).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Option to print progress. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>parallel</th>
      <td><p>Option to parallelise the process for obtaining predicted
probabilities. Must be number to indicate the number of cores to use in
parallel. Default is NULL.</p></td>
    </tr>
    <tr>
      <th>DBSCAN_params</th>
      <td><p>Parameters passed to <code>dbscan</code> in list form
containing 1) eps and 2) MinPts. If a vector of different values is passed to
either or both eps and MinPts, then each combination of these parameters is
evaluated to maximize normalized enrichment (NE) is the provided genomic 
annotations. Normalized Enrichment is calculated as the number of genomic 
annotations that overlap with flanked predicted boundary points (see the slope 
parameter) divided by the total number of predicted boundaries, averaged for 
all genomic annotations. Parameters yielding  maximum NE score are automatically 
selected for the final prediction. It is advisable to explore results 
of the NE test, available in the `preciseTADparams` slot of the returned 
object (NEmean - mean normalized enrichment, larger the better; k - number 
of PTBRs), to, potentially, find eps and MinPts parameters providing 
the number of PTBRs and the NE score better agreeing with the number
of boundaries used for training. Default: list(30000, 100). Required.</p></td>
    </tr>
    <tr>
      <th>slope</th>
      <td><p>Controls how much to flank the predicted TAD boundary points for
calculating normalized enrichment. Default: 5000 bases. Required.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>version of the human genome assembly. Used to filter out
bases overlapping centromeric regions. Accepted values - hg19 or 
hg38. Default: hg19</p></td>
    </tr>
    <tr>
      <th>BaseProbs</th>
      <td><p>Option to include the vector of probabilities for each 
base-level coordinate. Recommended to be used only when chromCoords is 
specified. Default: FALSE</p></td>
    </tr>
    <tr>
      <th>savetobed</th>
      <td><p>If true, preciseTAD regions (PTBRs) and preciseTAD points
(PTBPs) will be saved as BED-like files into the current folder 
(as.data.frame(GRanges)). File name convention: 
&lt;PTBRs/PTBPs&gt;_&lt;threshold&gt;_&lt;MinPts&gt;_&lt;eps&gt;.bed, e.g., PTBR_1_3_30000.bed. 
If multiple DBSCAN_params are specified, each result will
be saved in its own file. Default: FALSE</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list containing 4 elements including:
1) data frame with average (and standard deviation) normalized enrichment 
(NE) values for each combination of t and eps (only if multiple values are 
provided for at least paramenter; all subsequent summaries are applied to 
optimal combination of (t, eps)),
2) the genomic coordinates spanning each preciseTAD predicted region (PTBR),
3) the genomic coordinates of preciseTAD predicted boundaries points (PTBP),
4) a named list including summary statistics of the following:
PTBRWidth - PTBR width, PTBRCoverage - the proportion of bases within a PTBR
with probabilities that equal to or exceed the threshold (t=1 by default),
DistanceBetweenPTBR - the genomic distance between the end of the previous
PTBR and the start of the subsequent PTBR, NumSubRegions - the number of
the subregions in each PTBR cluster, SubRegionWidth - the width of
the subregion forming each PTBR, DistBetweenSubRegions -
the genomic distance between the end of the previous PTBR-specific subregion
and the start of the subsequent PTBR-specific subregion, NormilizedEnrichment
- the normalized enrichment of the genomic annotations used in the model 
around flanked PTBPs, and BaseProbs - a numeric vector of probabilities for 
each corresponding base coordinate.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Read in ARROWHEAD-called TADs at 5kb</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>arrowhead_gm12878_5kb</span><span class='op'>)</span>

<span class='co'># Extract unique boundaries</span>
<span class='va'>bounds.GR</span> <span class='op'>&lt;-</span> <span class='fu'><a href='extractBoundaries.html'>extractBoundaries</a></span><span class='op'>(</span>domains.mat <span class='op'>=</span> <span class='va'>arrowhead_gm12878_5kb</span>,
                               filter <span class='op'>=</span> <span class='cn'>FALSE</span>,
                               CHR <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CHR21"</span>, <span class='st'>"CHR22"</span><span class='op'>)</span>,
                               resolution <span class='op'>=</span> <span class='fl'>5000</span><span class='op'>)</span>

<span class='co'># Read in GRangesList of 26 TFBS and filter to include only CTCF, RAD21,</span>
<span class='co'>#SMC3, and ZNF143</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>tfbsList</span><span class='op'>)</span>

<span class='va'>tfbsList_filt</span> <span class='op'>&lt;-</span> <span class='va'>tfbsList</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>tfbsList</span><span class='op'>)</span> <span class='op'>%in%</span>
                                                 <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Gm12878-Ctcf-Broad"</span>,
                                                   <span class='st'>"Gm12878-Rad21-Haib"</span>,
                                                   <span class='st'>"Gm12878-Smc3-Sydh"</span>,
                                                   <span class='st'>"Gm12878-Znf143-Sydh"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>

<span class='co'># Create the binned data matrix for CHR1 (training) and CHR22 (testing)</span>
<span class='co'># using 5 kb binning, distance-type predictors from 4 TFBS from</span>
<span class='co'># the GM12878 cell line, and random under-sampling</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>123</span><span class='op'>)</span>
<span class='va'>tadData</span> <span class='op'>&lt;-</span> <span class='fu'><a href='createTADdata.html'>createTADdata</a></span><span class='op'>(</span>bounds.GR <span class='op'>=</span> <span class='va'>bounds.GR</span>,
                         resolution <span class='op'>=</span> <span class='fl'>5000</span>,
                         genomicElements.GR <span class='op'>=</span> <span class='va'>tfbsList_filt</span>,
                         featureType <span class='op'>=</span> <span class='st'>"distance"</span>,
                         resampling <span class='op'>=</span> <span class='st'>"rus"</span>,
                         trainCHR <span class='op'>=</span> <span class='st'>"CHR21"</span>,
                         predictCHR <span class='op'>=</span> <span class='st'>"CHR22"</span><span class='op'>)</span>

<span class='co'># Perform random forest using TADrandomForest by tuning mtry over 10 values</span>
<span class='co'># using 3-fold CV</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>123</span><span class='op'>)</span>
<span class='va'>tadModel</span> <span class='op'>&lt;-</span> <span class='fu'><a href='TADrandomForest.html'>TADrandomForest</a></span><span class='op'>(</span>trainData <span class='op'>=</span> <span class='va'>tadData</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,
                            testData <span class='op'>=</span> <span class='va'>tadData</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>,
                            tuneParams <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>mtry <span class='op'>=</span> <span class='fl'>2</span>,
                                            ntree <span class='op'>=</span> <span class='fl'>500</span>,
                                            nodesize <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>,
                            cvFolds <span class='op'>=</span> <span class='fl'>3</span>,
                            cvMetric <span class='op'>=</span> <span class='st'>"Accuracy"</span>,
                            verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
                            model <span class='op'>=</span> <span class='cn'>TRUE</span>,
                            importances <span class='op'>=</span> <span class='cn'>TRUE</span>,
                            impMeasure <span class='op'>=</span> <span class='st'>"MDA"</span>,
                            performances <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; + Fold1: mtry=2, ntree=500, nodesize=1 
#&gt; - Fold1: mtry=2, ntree=500, nodesize=1 
#&gt; + Fold2: mtry=2, ntree=500, nodesize=1 
#&gt; - Fold2: mtry=2, ntree=500, nodesize=1 
#&gt; + Fold3: mtry=2, ntree=500, nodesize=1 
#&gt; - Fold3: mtry=2, ntree=500, nodesize=1 
#&gt; Aggregating results
#&gt; Fitting final model on full training set</div><div class='input'>
<span class='co'># Apply preciseTAD on a specific 2mb section of CHR22:17000000-18000000</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>123</span><span class='op'>)</span>
<span class='va'>pt</span> <span class='op'>&lt;-</span> <span class='fu'>preciseTAD</span><span class='op'>(</span>genomicElements.GR <span class='op'>=</span> <span class='va'>tfbsList_filt</span>,
                 featureType <span class='op'>=</span> <span class='st'>"distance"</span>,
                 CHR <span class='op'>=</span> <span class='st'>"CHR22"</span>,
                 chromCoords <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fl'>17000000</span>, <span class='fl'>18000000</span><span class='op'>)</span>,
                 tadModel <span class='op'>=</span> <span class='va'>tadModel</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,
                 threshold <span class='op'>=</span> <span class='fl'>1.0</span>,
                 verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
                 parallel <span class='op'>=</span> <span class='cn'>NULL</span>,
                 DBSCAN_params <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1000</span>, <span class='fl'>10000</span>, <span class='fl'>30000</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>100</span>, <span class='fl'>1000</span><span class='op'>)</span><span class='op'>)</span>,
                 slope <span class='op'>=</span> <span class='fl'>5000</span>,
                 genome <span class='op'>=</span> <span class='st'>"hg19"</span>,
                 BaseProbs <span class='op'>=</span> <span class='cn'>FALSE</span>,
                 savetobed <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "Establishing bp resolution test data using a distance type feature space"
#&gt; [1] "Establishing probability vector"
#&gt; [1] "Determining optimal combination of epsilon neighborhood (eps) and minimum number of points (MinPts)"
#&gt; [1] "Initializing DBSCAN for MinPts=10 and eps=1000"
#&gt; [1] "    preciseTAD identified 12 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 12"
#&gt; [1] "        Cluster 2 out of 12"
#&gt; [1] "        Cluster 3 out of 12"
#&gt; [1] "        Cluster 4 out of 12"
#&gt; [1] "        Cluster 5 out of 12"
#&gt; [1] "        Cluster 6 out of 12"
#&gt; [1] "        Cluster 7 out of 12"
#&gt; [1] "        Cluster 8 out of 12"
#&gt; [1] "        Cluster 9 out of 12"
#&gt; [1] "        Cluster 10 out of 12"
#&gt; [1] "        Cluster 11 out of 12"
#&gt; [1] "        Cluster 12 out of 12"
#&gt; [1] "Initializing DBSCAN for MinPts=100 and eps=1000"
#&gt; [1] "    preciseTAD identified 12 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 12"
#&gt; [1] "        Cluster 2 out of 12"
#&gt; [1] "        Cluster 3 out of 12"
#&gt; [1] "        Cluster 4 out of 12"
#&gt; [1] "        Cluster 5 out of 12"
#&gt; [1] "        Cluster 6 out of 12"
#&gt; [1] "        Cluster 7 out of 12"
#&gt; [1] "        Cluster 8 out of 12"
#&gt; [1] "        Cluster 9 out of 12"
#&gt; [1] "        Cluster 10 out of 12"
#&gt; [1] "        Cluster 11 out of 12"
#&gt; [1] "        Cluster 12 out of 12"
#&gt; [1] "Initializing DBSCAN for MinPts=1000 and eps=1000"
#&gt; [1] "    preciseTAD identified 3 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 3"
#&gt; [1] "        Cluster 2 out of 3"
#&gt; [1] "        Cluster 3 out of 3"
#&gt; [1] "Initializing DBSCAN for MinPts=10 and eps=10000"
#&gt; [1] "    preciseTAD identified 5 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 5"
#&gt; [1] "        Cluster 2 out of 5"
#&gt; [1] "        Cluster 3 out of 5"
#&gt; [1] "        Cluster 4 out of 5"
#&gt; [1] "        Cluster 5 out of 5"
#&gt; [1] "Initializing DBSCAN for MinPts=100 and eps=10000"
#&gt; [1] "    preciseTAD identified 5 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 5"
#&gt; [1] "        Cluster 2 out of 5"
#&gt; [1] "        Cluster 3 out of 5"
#&gt; [1] "        Cluster 4 out of 5"
#&gt; [1] "        Cluster 5 out of 5"
#&gt; [1] "Initializing DBSCAN for MinPts=1000 and eps=10000"
#&gt; [1] "    preciseTAD identified 3 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 3"
#&gt; [1] "        Cluster 2 out of 3"
#&gt; [1] "        Cluster 3 out of 3"
#&gt; [1] "Initializing DBSCAN for MinPts=10 and eps=30000"
#&gt; [1] "    preciseTAD identified 4 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 4"
#&gt; [1] "        Cluster 2 out of 4"
#&gt; [1] "        Cluster 3 out of 4"
#&gt; [1] "        Cluster 4 out of 4"
#&gt; [1] "Initializing DBSCAN for MinPts=100 and eps=30000"
#&gt; [1] "    preciseTAD identified 4 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 4"
#&gt; [1] "        Cluster 2 out of 4"
#&gt; [1] "        Cluster 3 out of 4"
#&gt; [1] "        Cluster 4 out of 4"
#&gt; [1] "Initializing DBSCAN for MinPts=1000 and eps=30000"
#&gt; [1] "    preciseTAD identified 3 PTBRs"
#&gt; [1] "    Establishing PTBPs"
#&gt; [1] "        Cluster 1 out of 3"
#&gt; [1] "        Cluster 2 out of 3"
#&gt; [1] "        Cluster 3 out of 3"
#&gt; [1] "Optimal combination of MinPts and eps is = (10, 30000)"
#&gt; [1] "preciseTAD identified a total of 7799 base pairs whose predictive probability was equal to or exceeded a threshold of 1"
#&gt; [1] "Initializing DBSCAN for MinPts = 10 and eps = 30000"
#&gt; [1] "preciseTAD identified 4 PTBRs"
#&gt; [1] "Establishing PTBPs"
#&gt; [1] "Cluster 1 out of 4"
#&gt; [1] "Cluster 2 out of 4"
#&gt; [1] "Cluster 3 out of 4"
#&gt; [1] "Cluster 4 out of 4"</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Spiro Stilianoudakis, Mikhail Dozmorov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


